pkg/resource/deploy/providers/registry.go

```diff
@@ -466,20 +413,63 @@ func (r *Registry) Diff(urn resource.URN, id resource.ID, olds, news resource.Pr
 	return diff, nil
 }
 
-// Same executes as part of the "Same" step for a provider that has not changed. It exists solely to allow the registry
-// to point aliases for a provider to the proper object.
-func (r *Registry) Same(ref Reference) {
+// Same executes as part of the "Same" step for a provider that has not changed. It configures the provider
+// instance with the given state and fixes up aliases.
+func (r *Registry) Same(res *resource.State) error {
 	r.m.RLock()
 	defer r.m.RUnlock()
 
+	urn := res.URN
+	if !IsProviderType(urn.Type()) {
+		return fmt.Errorf("urn %v is not a provider type", urn)
+	}
+
+	// Ensure that this provider has a known ID.
+	if res.ID == "" || res.ID == UnknownID {
+		return fmt.Errorf("provider '%v' has an unknown ID", urn)
+	}
+
+	ref := mustNewReference(urn, res.ID)
 	logging.V(7).Infof("Same(%v)", ref)
 
-	// If this provider is aliased to a different old URN, make sure that it is present under both the old reference and
-	// the new reference.
+	if _, ok := r.providers[ref]; ok {
+		// Already configured
+		return nil
+	}
+
+	providerPkg := GetProviderPackage(urn.Type())
+
+	// Parse the provider version, then load, configure, and register the provider.
+	version, err := GetProviderVersion(res.Inputs)
+	if err != nil {
+		return fmt.Errorf("parse version for %v provider '%v': %v", providerPkg, urn, err)
+	}
+	downloadURL, err := GetProviderDownloadURL(res.Inputs)
+	if err != nil {
+		return fmt.Errorf("parse download URL for %v provider '%v': %v", providerPkg, urn, err)
+	}
+	// TODO: We should thread checksums through here.
+	provider, err := loadProvider(providerPkg, version, downloadURL, nil, r.host, r.builtins)
+	if err != nil {
+		return fmt.Errorf("load plugin for %v provider '%v': %v", providerPkg, urn, err)
+	}
+	if provider == nil {
+		return fmt.Errorf("find plugin for %v provider '%v' at version %v", providerPkg, urn, version)
+	}
+	if err := provider.Configure(res.Inputs); err != nil {
+		closeErr := r.host.CloseProvider(provider)
+		contract.IgnoreError(closeErr)
+		return fmt.Errorf("configure provider '%v': %v", urn, err)
+	}
+
+	logging.V(7).Infof("loaded provider %v", ref)
+	r.providers[ref] = provider
+
 	if alias, ok := r.aliases[ref.URN()]; ok {
-		aliasRef := mustNewReference(alias, ref.ID())
-		r.providers[ref] = r.providers[aliasRef]
+		r.providers[mustNewReference(alias, ref.ID())] = provider
```

Comment:
We only have a read lock here. We cannot mutate. Need to switch to a write lock.
As optimization for the common case, we should keep the read lock up until the check-map-and-early-return.
After that, unlock, get a write lock, check again (in case another goroutine was also trying to create that provider), and if not, do the rest of this.

Reply:
Ah good catch

Reply:
Interestingly this was previously unlocked but I guess hit rarely enough that we never trigger a concurrent write bug.